R/register_handlers.R

```diff
@@ -0,0 +1,98 @@
+#' Register handlers for logging messages, warnings and errors
+#'
+#' @param namespace (`character(1)`) the logger namespace
+#' @param package (`character(1)`) the package name
+#'
+#' @return `NULL` invisibly. Called for its side effects.
+#'
+#' @details This function registers global handlers for messages, warnings and errors.
+#' The handlers will investigate the call stack and if it contains a function
+#' from the package, the message, warning or error will be logged into the respective
+#' namespace.
+#' The handlers are registered only once per package and type. Consecutive calls will be ignored.
+#'
+#' @note Registering handlers is disallowed within `tryCatch` blocks which is used throughout
+#' base R and other packages. Because of this, the `register_handlers` function is wrapped in a
+#' `try` block to avoid errors.
+#'
+#' @seealso [globalCallingHandlers()]
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' register_logger(namespace = "teal.logger")
+#' register_handlers("teal.logger")
+#' # see the outcome
+#' globalCallingHandlers()
+#' }
+register_handlers <- function(namespace, package = namespace) {
+  try(
+    {
+      register_handler_type(namespace = namespace, package = package, type = "message")
+      register_handler_type(namespace = namespace, package = package, type = "warning")
+      register_handler_type(namespace = namespace, package = package, type = "error")
+    },
+    silent = TRUE
```

Comment:
@pawelru what about a `tryCatch` with a message on a failure, so that it's visible to the user that logging failed? Unsure who is really keen on seeing logs, but a message on a failure does not hurt.

Reply:
I tried this - forgot to mention it sorry.
The problem with it is that this `try` is meant for R CMD CHECK, test, pkgdown and all others that does not accept changing handlers. When I throw a warning then R CMD BUILD failed. Probably something else will do as well. I haven't tried with the message though - I just abandoned this after failing with a warning.
That's the story behind `try` with `silent = TRUE`.

So now the question is only about a message. I'm thinking whether it is useful and I honestly don't know. It would be good to send that information (somewhere, somehow) but on the other hand I think I don't want to clobber console a lot (or: at all). I don't know...

Reply:
I think that a message is doable. Will implement it soon

Reply:
Tested locally with a message. 

Impact:
- R CMD BUILD - everything ok - no impact
- R CMD CHECK - a note:
```
── R CMD check results ──────────────────────────────────────────────────────────────────────────── teal.logger 0.1.3.9011 ────
Duration: 10.7s

❯ checking startup messages can be suppressed ... NOTE
  Failed to register handlers
  
  It looks like this package (or a package it requires) has a startup
  message which cannot be suppressed: see ?packageStartupMessage.

0 errors ✔ | 0 warnings ✔ | 1 note ✖
```
- test - a new entry in the console
```
r$> devtools::test()
ℹ Testing teal.logger
Failed to register handlers          # <- HERE!!!!!!!!!
✔ | F W  S  OK | Context
✔ |         10 | register_logger                                   

══ Results ════════════════════════════════════════════════════════
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 10 ]
```

- pkgdown - a new entry in the log
```
r$> pkgdown::build_site()
-- Installing package into temporary library ----------------------
== Building pkgdown site =======================================================
Reading from: '/Users/ruckip/Documents/repo/gh/insightsengineering/teal.logger'
Writing to:   '/Users/ruckip/Documents/repo/gh/insightsengineering/teal.logger/docs'
-- Initialising site -----------------------------------------------------------
Copying '../../../../../Library/R/arm64/4.3/library/pkgdown/BS5/assets/link.svg' to 'link.svg'
Copying '../../../../../Library/R/arm64/4.3/library/pkgdown/BS5/assets/pkgdown.js' to 'pkgdown.js'
Copying '../../../../../Library/R/arm64/4.3/library/nesttemplate/pkgdown/assets/analytics.js' to 'analytics.js'
Copying '../../../../../Library/R/arm64/4.3/library/nesttemplate/pkgdown/assets/consent.css' to 'consent.css'
Copying '../../../../../Library/R/arm64/4.3/library/nesttemplate/pkgdown/assets/consent.js' to 'consent.js'
Copying '../../../../../Library/R/arm64/4.3/library/nesttemplate/pkgdown/assets/cookie_policy.txt' to 'cookie_policy.txt'
-- Building home ---------------------------------------------------------------
Writing 'authors.html'
Reading 'SECURITY.md'
Writing 'SECURITY.html'
Reading '.github/CODE_OF_CONDUCT.md'
Writing 'CODE_OF_CONDUCT.html'
Reading '.github/CONTRIBUTING.md'
Writing 'CONTRIBUTING.html'
Reading '.github/pull_request_template.md'
Writing 'pull_request_template.html'
Writing 'LICENSE-text.html'
Failed to register handlers          # <- HERE!!!!!!!!!
Copying 'man/figures/lifecycle-archived.svg' to 'reference/figures/lifecycle-archived.svg'
Copying 'man/figures/lifecycle-defunct.svg' to 'reference/figures/lifecycle-defunct.svg'
Copying 'man/figures/lifecycle-deprecated.svg' to 'reference/figures/lifecycle-deprecated.svg'
Copying 'man/figures/lifecycle-experimental.svg' to 'reference/figures/lifecycle-experimental.svg'
Copying 'man/figures/lifecycle-maturing.svg' to 'reference/figures/lifecycle-maturing.svg'
Copying 'man/figures/lifecycle-questioning.svg' to 'reference/figures/lifecycle-questioning.svg'
Copying 'man/figures/lifecycle-soft-deprecated.svg' to 'reference/figures/lifecycle-soft-deprecated.svg'
Copying 'man/figures/lifecycle-stable.svg' to 'reference/figures/lifecycle-stable.svg'
Copying 'man/figures/lifecycle-superseded.svg' to 'reference/figures/lifecycle-superseded.svg'
Writing '404.html'
-- Building function reference -------------------------------------------------
Writing 'reference/index.html'
Reading 'man/layout_teal_glue_generator.Rd'
Writing 'reference/layout_teal_glue_generator.html'
Reading 'man/log_system_info.Rd'
Writing 'reference/log_system_info.html'
Reading 'man/register_handlers.Rd'
Writing 'reference/register_handlers.html'
Reading 'man/register_logger.Rd'
Writing 'reference/register_logger.html'
Reading 'man/suppress_logs.Rd'
Writing 'reference/suppress_logs.html'
Reading 'man/teal.logger-package.Rd'
Writing 'reference/teal.logger-package.html'
-- Building articles -----------------------------------------------------------
Writing 'articles/index.html'
Reading 'vignettes/teal-logger.Rmd'
Writing 'articles/teal-logger.html'
-- Building news ---------------------------------------------------------------
Writing 'news/index.html'
Writing 'sitemap.xml'
-- Building search index -------------------------------------------------------
== DONE ========================================================================
-- Previewing site ------------------------------------------------
```

I think that this producing a note disqualifies a message :( Converting this into a `packageStartupMessage` is not necessarily correct as this theoretically could be registered outside of package startup. 
What do you think? (haven't pushed this yet)

Reply:
Yeah, I dislike the note. I wonder if there is a way to have a check that detects if a R CMD CHECK is run, and if yes, the we could disable the message for that situation. But I don't recall such functionality.

Why this can not happen during R CMD CHECK? If you know that, maybe you can create a if+skip statement for this situation, so it does not popup during R CMD CHECK?

Reply:
I think that with some effort I would be able to came up with a condition whether or not we are during R CMD CHECK but I think that I don't want to go with that route. Let me explain why. If we have it for R CMD CHECK then why not `devtools::test()` and then `pkgdown::build_site()`. The problem is that such list is open and unknown. It would be hard to have a complete list of cases. Not to mention that each of these could bring a non-obvious and not really required dependencies (just a guess - in the package namespace we might have: `pkgdown::check_if_inside_render()` - a `pkgdown` dependency is not really needed for functionalities here).

That's why I think that try / tryCatch is more robust even though not that accurate.

Reply:
But it's not really about the R CMD CHECK but rather becuase there are circumstances inside R CMD CHECK. What cause them? Is it some lack of privileges? Non-interactive execution? Special setup of ENVIRONMENTAL VARIABLES used for R CMD CHECK? If we knew the direct cause, maybe it would be easier to understand and provide a meaningful skip condition. For now I only get that something blocks registration during R CMD CHECK and we use `try` which will not inform us about failures that happen outside of R CMD CHECK.

Reply:
ok I know what you are referring to - you are trying to find a

I have to admit that this is still not super clear for me. See my comment here: https://github.com/insightsengineering/teal.logger/pull/73#issuecomment-1956691558

I managed to find only one resource which still does not provide enough information. It looks like it's about registering handlers within a try-block. However, I was not able to confirm or deny. Maybe I should come back to this... Please join me if you can - I might need some help on this.

Reply:
Actually the `try` - based solution is not working at all :( Sorry for bothering you with a review. 

I came back to this and as expected - this is not easy topic. I haven't found a way to get the current handlers stack (not to be confused with global handlers stack). The error is coming from the internals of R.
I have explored again the idea of block-list and to me there are essentially only two functions that we need to block - `tryCatch()` (with `try` being a simple wrapper) and `withCallingHandlers()` (with its set of wrappers). I hope that this fulfils all of the use case. I have ran some tests and it's good so far.

Reply:
Just for the record - it's coming from this: https://github.com/wch/r-source/blob/e790c6820634299cfd62491f24255de89b940676/src/main/errors.c#L2588-L2615

I don't have access to the variables used :(

Reply:
I think I'm out of the loop in here, but glad to hear you know it's the try that prevents the stuff from working as expected